File: LeetCodePractice/src/_0004_Stack/_0007_LargestRectangleInHistogram.java
package _0004_Stack;

import java.util.Stack;

public class _0007_LargestRectangleInHistogram {

//	Solution from AlgorithmMadeEasy Youtube
	public static int largestRectangleArea(int[] h) {
		int n = h.length;
		int maxArea = 0;
		Stack<Integer> s = new Stack();
		for (int i = 0; i <= n; i++) {

			int currHeight = i == n ? 0 : h[i];
			// currHeight › h[top] ? push(i) : pop & find area
			while (!s.isEmpty() && currHeight < h[s.peek()]) {
				int top = s.pop();
				int width = s.isEmpty() ? i : i - s.peek() - 1;
				int area = h[top] * width;
				maxArea = Math.max(area, maxArea);
			}
			s.push(i);
		}
		return maxArea;
	}

	public static void main(String args[]) {
		System.out.println(largestRectangleArea(new int[] { 2, 1, 5, 6, 2, 3 }));
	}
}

File: LeetCodePython/_0004_Stack/_0007_LargestRectangleInHistogram.py
from typing import List

# Solution from AlgorithmMadeEasy Youtube
# HARD QUESTION:
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        max_area = 0
        stack = []
    
        for i in range(n + 1):
            curr_height = 0 if i == n else heights[i]
        
            while stack and curr_height < heights[stack[-1]]:
                top = stack.pop()
                width = i if not stack else i - stack[-1] - 1
                area = heights[top] * width
                max_area = max(area, max_area)
        
            stack.append(i)
    
        return max_area
    
sol = Solution()
print(sol.largestRectangleArea([2, 1, 5, 6, 2, 3]))